File: java/src/main/java/demo/officeDay/generics/AbstractServiceCommunicator.java
package demo.officeDay.generics;

import java.lang.reflect.ParameterizedType;

public abstract class AbstractServiceCommunicator<IN, OUT> {

    public OUT communicate(IN input) throws IllegalAccessException, InstantiationException, ClassNotFoundException {
        System.out.println(getServiceByName(input));

        OUT response = getResponseClass().newInstance();
        System.out.println("RESPONSE: " + response.getClass().getName());

        return response;
    }

    private String getServiceByName(IN input) {
        return "GET /" + input.getClass().getSimpleName();
    }

    protected abstract Class<OUT> getResponseClass();

    private Class<OUT> getResponseClassReflection() throws ClassNotFoundException {
        String genericResponseClassName = ((ParameterizedType) getClass().getGenericSuperclass()).getActualTypeArguments()[1].getTypeName();
        return (Class<OUT>) Thread.currentThread().getContextClassLoader().loadClass(genericResponseClassName);
    }
}

File: csharp/demo.officeDay/generics/AbstractServiceCommunicator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace demo.officeDay.generics
{
    public class AbstractServiceCommunicator<IN, OUT>
    {

        public OUT communicate(IN input)
        {
            System.Console.WriteLine(getServiceByName(input));

            OUT response = System.Activator.CreateInstance<OUT>();
            System.Console.WriteLine("RESPONSE: " + response.GetType().FullName);

            return response;
        }

        private String getServiceByName(IN input)
        {
            return "GET /" + input.GetType().Name;
        }

    }
}
